File: ProblemaF/src/F.cpp
#include <bits/stdc++.h>
using namespace std;

// Caché: altura del árbol de raíz i
vector<int> height;

// Caché: ¿es robusto el árbol de raíz i?
vector<optional<bool>> robust;

// Caché: ¿un fallo en una sola hoja disminuiría la altura
// del árbol de raíz i?
vector<optional<bool>> can_shorten;

vector<vector<int>> adj;

int getHeight(int node) {
	if (height[node] != -1) return height[node];
	if (adj[node].empty()) {
		height[node] = 0;
		return 0;
	}

	int thisHeight = getHeight(adj[node][0]);
	if (adj[node].size() > 1) thisHeight = max(thisHeight, getHeight(adj[node][1]));
	height[node] = 1 + thisHeight;

	return height[node];
}

bool canShorten(int node) {
	if (can_shorten[node].has_value()) return can_shorten[node].value();

	// Si un nodo es hoja y raíz, su altura pasa de 1 a 0 si ella falla
	if (adj[node].empty()) {
		can_shorten[node] = true;
		return true;
	}

	if (adj[node].size() == 1) {
		// Si un nodo solo tiene un hijo, será acortable si su hijo es acortable
		can_shorten[node] = canShorten(adj[node][0]);
		return can_shorten[node].value();
	}

	// Si un nodo tiene dos hijos, será acortable si su altura proviene de uno
	// de los hijos y ese hijo es acortable
	int left = adj[node][0];
	int right = adj[node][1];
	int leftHeight = getHeight(left);
	int rightHeight = getHeight(right);

	if (leftHeight == rightHeight) {
		// Ambos subárboles tienen la misma altura. Incluso si uno pudiese
		// acortarse, nunca lo harán los dos a la vez
		can_shorten[node] = false;
		return false;
	}

	can_shorten[node] =(leftHeight > rightHeight && canShorten(left)) || (rightHeight > leftHeight && canShorten(right)) ;
	return can_shorten[node].value();
}

bool isRobust(int node) {
	if (robust[node].has_value()) return robust[node].value();

	for (int child : adj[node]) {
		if (!isRobust(child)) {
			robust[node] = false;
			return false;
		}
	}

	if (adj[node].empty()) {
		// Nodo hoja siempre es robusto
		robust[node] = true;
		return true;
	}

	auto [leftHeight, leftShortable] = make_tuple(getHeight(adj[node][0]), canShorten(adj[node][0]));

	auto [rightHeight, rightShortable] = adj[node].size() > 1
		? make_tuple(getHeight(adj[node][1]), canShorten(adj[node][1]))
		: make_tuple(0, false);

	auto [shortestHeight, shortestShortable] = leftHeight < rightHeight
		? make_tuple(leftHeight, leftShortable)
		: make_tuple(rightHeight, rightShortable);

	auto [tallestHeight, tallestShortable] = leftHeight < rightHeight
		? make_tuple(rightHeight, rightShortable)
		: make_tuple(leftHeight, leftShortable);

	if (shortestHeight == tallestHeight) {
		// Si los dos subárboles tienen la misma altura, incluso si uno pierde
		// altura van a seguir equilibrados
		robust[node] = true;
		return true;
	}

	if (shortestHeight < (tallestHeight - 1) || shortestShortable) {
		// Si los subárboles ya tienen una diferencia mayor que uno (no debería ocurrir)
		// o tienen una diferencia de uno pero el más corto puede acortarse si una
		// hoja fallase, la diferencia sería de dos y no serían robustos
		robust[node] = false;
		return false;
	}

	robust[node] = true;
	return true;
}

void solve() {
	int N;
	cin >> N;

	adj = vector<vector<int>>(N);
	height = vector<int>(N, -1);
	robust = vector<optional<bool>>(N, nullopt);
	can_shorten = vector<optional<bool>>(N, nullopt);

	for (int i = 0; i < N; ++i) {
		int k;
		cin >> k;
		for (int j = 0; j < k; ++j) {
			int child;
			cin >> child;
			adj[i].push_back(child-1);
		}
	}

	if(isRobust(0)) cout << "SI\n";
	else            cout << "NO\n";
}

int main() {
	int T;
	cin >> T;
	while (T--) {
		solve();
	}
}

File: ProblemaF/src/F.py
# Un castillo es un árbol binario que cuelga del estallido 1. La "altura" de un
# subárbol (hojas = 0) determina la duración de su subcastillo (duración = altura + 1),
# así que trabajamos directamente con alturas.
#
# El castillo es robusto si sigue siendo armonioso tras el fallo de cualquier hoja.
# Quitar una hoja reduce la altura de un subárbol en 0 o en 1, nunca más. Definimos:
#   height(v)      -> altura del subárbol de raíz v.
#   can_shorten(v) -> ¿existe una hoja cuyo fallo reduzca height(v)?
#   is_robust(v)   -> ¿el subárbol de raíz v aguanta armonioso cualquier fallo de hoja?
#
# Como el árbol es armonioso (hermanos con alturas que difieren <= 1) y los nodos de un
# solo hijo cuelgan siempre de una hoja, el árbol está equilibrado al estilo AVL: su
# altura es O(log N), de modo que la recursión no se acerca al límite de Python.

import sys
from functools import lru_cache


def solve(children):
	@lru_cache(maxsize=None)
	def height(v):
		ch = children[v]
		if not ch:
			return 0
		return 1 + max(height(c) for c in ch)

	@lru_cache(maxsize=None)
	def can_shorten(v):
		ch = children[v]
		if not ch:
			# Una hoja desaparece: su altura pasa de 0 a "nada", el padre la nota.
			return True
		if len(ch) == 1:
			# El único hijo es una hoja; el nodo se acorta si su hijo se acorta.
			return can_shorten(ch[0])
		left, right = ch
		hl, hr = height(left), height(right)
		if hl == hr:
			# La altura viene de ambos hijos: ningún fallo aislado la reduce.
			return False
		# La altura viene solo del hijo más alto: se acorta si ese se acorta.
		return can_shorten(left) if hl > hr else can_shorten(right)

	@lru_cache(maxsize=None)
	def is_robust(v):
		ch = children[v]
		# Todo subcastillo interno debe aguantar sus propios fallos.
		for c in ch:
			if not is_robust(c):
				return False
		if not ch:
			return True

		hl, sl = height(ch[0]), can_shorten(ch[0])
		if len(ch) > 1:
			hr, sr = height(ch[1]), can_shorten(ch[1])
		else:
			# Un solo hijo (hoja): al fallar, el nodo queda como estrella final. Sano.
			hr, sr = 0, False

		# Nos interesa el hijo más bajo: es el que rompe la armonía si se acorta.
		if hl <= hr:
			short_h, short_can, tall_h = hl, sl, hr
		else:
			short_h, short_can, tall_h = hr, sr, hl

		if short_h == tall_h:
			# Alturas iguales: aunque una baje en 1, la diferencia queda en 1. Armonioso.
			return True
		# Difieren en 1: si el más bajo puede encoger, la diferencia sube a 2. Roto.
		return not short_can

	return is_robust(0)


def main():
	data = sys.stdin.buffer.read().split()
	pos = 0
	T = int(data[pos]); pos += 1
	out = []
	for _ in range(T):
		N = int(data[pos]); pos += 1
		children = [None] * N
		for i in range(N):
			k = int(data[pos]); pos += 1
			children[i] = [int(data[pos + j]) - 1 for j in range(k)]
			pos += k
		out.append("SI" if solve(children) else "NO")
	sys.stdout.write("".join(line + "\n" for line in out))


main()
